contents/blog/cypress-end-to-end-tests.md

```diff
@@ -0,0 +1,229 @@
+
+### Unit & Integration tests
+
+In our Django codebase you'll find good number of [Django tests](https://github.com/PostHog/posthog/tree/master/posthog/test) that help keep us honest as we hack away at the backend of Posthog that keeps track of all the 1's and 0's that our customers depend on for making product decisions. These are our frontline defenders that let us know that something might be up before we even get to the point of creating a PR.
+
+For this we do lean heavily on the standard Django test runner.
+
+If you are interested in learning more on testing with Django check out Django's great [docs](https://docs.djangoproject.com/en/3.0/topics/testing/) on testing.
+
+These tests only get you so far though. You know that the backend is going to behave well after you land the changes that you've made, but what if you accidently changed something that breaks the frontend in weird and unexpected ways?
+
+### Enter Cypress
+
+According to Cypress's Github [repo](https://github.com/cypress-io/cypress) it is a **fast, easy and reliable testing for anything that runs in a browser**. What does that mean exactly though?
+
+It lets you programmatically interact with your application by querying the DOM and running actions against any selected elements. You can see that in a few of our [Cypress test definitions](https://github.com/PostHog/posthog/tree/master/cypress/integration)
+
+#### Example of our integration test for our Funnel user experience:
+
+```js
+describe('Funnels', () => {
+  //boilerplate to make sure we are on the funnel page of the app
+  beforeEach(() => {
+    cy.get('[data-attr=menu-item-funnels]').click()
+  })
+
+  // Test to make sure that the funnel page actually loaded
+  it('Funnels loaded', () => {
+    cy.get('h1').should('contain', 'Funnels')
+  })
+
+  // Test that when we select a funnel then we can edit that funnel
+  it('Click on a funnel', () => {
+    cy.get('[data-attr=funnel-link-0]').click()
+    cy.get('[data-attr=funnel-editor]').should('exist')
+  })
+
+  // Test that we can create a new funnel when we click 'create funnel' button
+  it('Go to new funnel screen', () => {
+    cy.get('[data-attr=create-funnel]').click()
+    cy.get('[data-attr=funnel-editor]').should('exist')
+  })
+
+  // Test that we can create a new funnel end to end
+  it('Add 1 action to funnel', () => {
+    cy.get('[data-attr=create-funnel]').click()
+    cy.get('[data-attr=funnel-editor]').should('exist')
+
+    cy.get('[data-attr=edit-funnel-input]').type('Test funnel')
+    cy.get('[data-attr=add-action-event-button]').click()
+    cy.get('[data-attr=trend-element-subject-0]').click()
+    cy.contains('Pageviews').click()
+
+    cy.get('[data-attr=save-funnel-button]').click()
+
+    cy.get('[data-attr=funnel-viz]').should('exist')
+  })
+})
```

Comment:
Cool post!

I'd add a few lines here as well about `data-attr`. I'm not sure how industry-wide this practice is, but adding these kinds of "hint attributes" into the dom makes for sturdier and cleaner tests.

Reply:
@fuziontech Here's a snippet you can bring paste in. I avoided doing so myself in case you had diffs to commit I don't want to disrupt.

#### Tracking the elements
To keep our tested elements clear, manageable, and reusable upon refactor, we take advantage of the element attributes that html and react specifically [recognize](https://reactjs.org/blog/2017/09/08/dom-attributes-in-react-16.html). Cypress has an amazing built in inspector on their test-runner that allows you to identify elements that you would like to add tests to. While the tool works great, we found that occasionally the heavily nested components and classes would create selectors that were inflexible. With the `data-attr` tag, we just need to keep track of the tag when updating/changing the components we're using without needing to rely on the inspector to find the precise selector for the test!
```
<LineGraph
  data-attr="trend-line-graph"
  ...
/>
```

Reply:
Added